### ISSUE TITLE
Assignment of backup instances

### ISSUE THREAD

**Issue by wvanmaris** [bitfocus/companion#268]
**Describe the feature**
The possibility to assign a backup instance to a main instance in instance settings. E.g. you have a QLAB main system and want to control a QLAB backup system. All buttons can be made with just a command for the QLAB main system, but because the backup system is assigned as backup instance for the main, all commands are sent to both.
This can be passed by, by just adding cues for both instances on each button, but by having a general "follow" / "clone" / "backup" option, it will save a lot of work.
Perhaps the feature can be further enhanced by being able to send a delay option for the backup command.

**Is this platform dependent (windows, mac, ..)?**
no

**If documentation is required to implement, do you know where to find it?**
If applicable, add screenshots, protocol description PDFs, etc to help the developers along

**Usecases**
Useful for any system that often is backed by a secondary backup system (mediaservers / lighting consoles / seamless switchers, or even (if multiple backup instances could be applied) projectors and such (for global shutter commands).

**Comment by skermanator** [comment]
Would love this feature

**Comment by sphlabs** [comment]
Just bumping this to see if there are any active plans/ideas on implementing this feature?

For more complex setups, this would be an immense time/error saver, vs. manually duplicating all actions on all buttons.

The main complication I can see is in handling feedback and variables (although feedback might not be so bad in a lot of cases) but at least having this for actions would be a big win.

**Comment by Julusian** [comment]
I see a couple of challenges which need good solutions for this:

Different connections from the same module can have different actions/feedbacks exposed  
Different connections from the same module can have different options for an action/feedback (while the name of something might be the same, it could have a different id internally.  
This means that a companion-level primary/backup setup wont work with some modules. I dont see a good way around that.

Building on that, what if the systems are setup slightly differently, and need slightly different parameters? perhaps inputs to the mixer are patched slightly differently for reasons out of your control

Do we need a way to add actions that address the 'pool' as well as an option to address an individual connection in that pool?

Which do we use for feedback?  
I am thinking that there should be a panel to select which is the primary, and it is the primary that we use for feedback and variables. To begin with we could keep the primary to be manually selected, that could be expanded to auto select one later on

I dont really have any experience in this area of using backups for this kind of equipment, so I am not entirely sure if these concerns are valid, or how best to solve them.
As such, I will find it hard to motivate spending my time on this instead of other features. It may happen, but it definitely needs more of a plan on how it should work before I will consider it

**Comment by Sullyy** [comment]
"Useful for any system that often is backed by a secondary backup system (media servers / lighting consoles / seamless switchers, or even (if multiple backup instances could be applied) projectors and such (for global shutter commands)."
At least 3 from this examples media server, seamless switchers - projectors, if not all4 need to be idemntical and with identical connection, setup, programing, to function as backup anyway, so the feedback would be the only issue here i think. I would mainly use this feature for media server, resolume, millumin... witch i backup when i can, and use identical machines. for this feature I would easy give up feedback, or use an internal feedback - button pressed or something just to show me where i am, i would take a-lot of time off preparing a show

**Comment by Sullyy** [comment]
or is this "the possibility copy the commands sent to an instance to another similar instance. ex: 2 resolume, 1 is backup- i need to copy all the commands from the main to the backup resolume." from my thread witch you closed earlier as it was a duplicate, an easier solution?

**Comment by sphlabs** [comment]
I would echo that sentiment. The idea would be to send identical commands to the backup device; the idea being that they are identically configured and should be running in lock-step with the primary. For setups where the backup system is doing something different, then that would have to be programmed manually as a second device.

But there are many, many production that have identical systems running in tandem for which this would be applicable.

In terms of feedback/variables: If instances are flagged as primary/backup (or primary/secondary/tertiary/etc) then feedback could come from the primary system. To be fancy/stretch goal--if the primary loses connection then feedback comes from secondary, etc. This could include cases where the primary is manually disabled via a button for cases where it's not working properly but still connected.

Octopus does this, or something very similar; although it's a while since I've used it so can't remember all the specifics. But it does have a primary/backup concept, and can support multiple backups. They all get the exact same commands as the primary.

**Comment by Julusian** [comment]
Yeah the initial implementation could definitely make the assumption that everything is identical.
But I worry that 'they are identically configured' can be misleading.
I know that the vmix api uses guids for a bunch of things, as global consistent identifiers for things. So while you will probably start off by building a project file and copying it to the backup machine, when you get to adding an extra thing last minute, you would probably rather follow the same steps on each machine to add the missing thing, resulting in them having different ids.
Again, this doesnt have to be a solved for the first iteration, but I do expect some frustration from users over this limitation if its not communicated well

**Comment by rywolf** [comment]
Yes this please!!!!!
Backup instances that do what the primary does.

**Comment by kman1898** [comment]
Any updates on this?

**Comment by sphlabs** [comment]
To come back to @Julusian ’s concerns from his last post:

I think assuming things are identical is the end assumption, not just the initial one, if things aren’t identical then the programmer doesn’t have a primary/backup setup and it isn’t Companion’s place to make decisions in that circumstance.

I’m the example of vMix, I wonder how many people use guids rather than the input name or index? I suppose with appropriate hooks, a module could be aware of backups, and a primary using guids could somehow communicate alternate means of reference…? But that’s really getting into the weeds and over complicating things for the majority of cases.

In the spirit of “don’t let perfection be the enemy of good enough”, I do think an initial implementation with some “here be dragons” warning would make a lot of people very happy!

Oh, and I do like the idea I read in a duplicate thread of specifying an instance-level delay time for each backup instance.

**Comment by wvanmaris** [comment]
Regarding Stephen Harrison: I see you point, but that was not the thought behind it. We’re talking a true backup where it NEEDS to be the same. Like a HOT backup ppt machine or QLAB instance. In real life you would edit on one machine, sync all content to the second and just need a way of controlling them both without having to enter two command on a single companion button. I think you’re overthinking the request. And yes, having a possibility to delay the command to the backup instance it valid. Walter van MarisMobiel verstuurdOp 15 mrt. 2023 om 22:24 heeft Stephen Harrison ***@***.***> het volgende geschreven:﻿
To come back to @Julusian ’s concerns from his last post:
I think assuming things are identical is the end assumption, not just the initial one, if things aren’t identical then the programmer doesn’t have a primary/backup setup and it isn’t Companion’s place to make decisions in that circumstance.
I’m the example of vMix, I wonder how many people use guids rather than the input name or index? I suppose with appropriate hooks, a module could be aware of backups, and a primary using guids could somehow communicate alternate means of reference…? But that’s really getting into the weeds and over complicating things for the majority of cases.
In the spirit of “don’t let perfection be the enemy of good enough”, I do think an initial implementation with some “here be dragons” warning would make a lot of people very happy!
Oh, and I do like the idea I read in a duplicate thread of specifying an instance-level delay time for each backup instance.

**Comment by sphlabs** [comment]
On the contrary, I agree with you completely, that the backup systems do need to be identically configured, and that that is the user's responsibility, not Companion's.

**Comment by Julusian** [comment]
ok, it sounds like there is a plan formed here now then.

one question I had hasnt been answered
> Do we need a way to add actions that address the 'pool' as well as an option to address an individual connection in that pool?

I'm thinking this as a case of you might intentionally want to unsync them at some point, so you can check your routing or something. for example, set the primary to red and backup to blue.  
this could be handled by adding each device twice, once as part of the primary/backup system, and once as individual connections. so maybe this can be ignored? 

It sounds like what this needs now is someone with time to figure out a how to implement it and then do it

**Comment by wvanmaris** [comment]
The way I envisioned it, was two separate instances, and having the option at any of them to assign the other as backup. With the condition it needs to be the same type of instance. In this backup instance setting it would be great to have a general delay setting or the likes. And having feedback from both would be awesome (as in a orange square if one of the instances is not available and red if both are. Walter van MarisMobiel verstuurdOp 17 mrt. 2023 om 17:38 heeft Julian Waller ***@***.***> het volgende geschreven:﻿
ok, it sounds like there is a plan formed here now then.
one question I had hasnt been answered

Do we need a way to add actions that address the 'pool' as well as an option to address an individual connection in that pool?

I'm thinking this as a case of you might intentionally want to unsync them at some point, so you can check your routing or something. for example, set the primary to red and backup to blue.
this could be handled by adding each device twice, once as part of the primary/backup system, and once as individual connections. so maybe this can be ignored?
It sounds like what this needs now is someone with time to figure out a how to implement it and then do it

**Comment by kman1898** [comment]
Any new progress in this realm?

**Comment by Federico-Catelotti** [comment]
@Julusian Thanks for referring me to this issue. I had already made an issue years ago for this in the vMix module but this change is increasingly important so a few weeks ago I opened another general issue and they brought me back here.

An option like this is essential, not only for a single module but for a tons of module. Honestly now when I have to do this I use central control. Joe Demax made a "module" to which you can not assign actions but only copies the actions of a “module” and replicates them on an unlimited different ip. it is helping me a lot because it speeds up the work. for all those who are here looking for a solution, until it is implemented in companion I like to use central control which solves the problem.

**Comment by Federico-Catelotti** [comment]
An option that lets you choose which of the identical modules to take feedback from might solve the feedback problem.

I imagine that in a module you have to put as many IPs as you want and choose which one is the master and which are the slaves. Once this is done, the feedback follows the master. If you want, you can make a trigger that, if it detects the disconnection of the master, moves the selection of the IP from which to take the feedback to an IP of your choice among the slaves.

**Comment by thedist** [comment]
I can't speak for other modules, but for the modules I develop such as vMix it's not particularly a priority feature to invest time developing as productions that are at the scale to have redundant vMix machines usually have an engineer that make their own solution in-house, or pay to hire a developer to make a proxy, so the limited number of users that would need this implemented in a free product like Companion has always been very limited.

**Comment by kman1898** [comment]
Yes but that’s why you wouldn’t develop it on a per module basis. You’d do it globally like many have asked for. It’s just the exact way it’s implemented that’s up for interpretation. But it would save so much time and effort on show site. 

Not sure what you mean by limited number of users. This would be widely used. Anyone that already uses companion in the AV world with redundant machines would find this useful and extremely time saving.

**Comment by thedist** [comment]
> Yes but that’s why you wouldn’t develop it on a per module basis. You’d do it globally like many have asked for.

Because it's something that would have to be done on a per module basis to be done correctly. Even if on Companions side you could designate a connection as a backup connection, that connection would still be trying to perform actions like parsing data, generating variables, things like that which have an processing overhead, so it would needlessly waste resources.

Additionally there would be no way to do consistency checking, which using vMix as an example when using a proxy to have redundant vMix machines you would also be validating the state of both vMix instances, ensuring the have the same inputs with the same Name/Number/GUIDs as well as the state of those inputs, audio routing and volumes etc... So if for some reason the 2 vMix machines get out of sync there would be absolutely no way to know that if this 'backup instance' concept was handled on Companions side rather than in individual modules. It would be negligent to run a professional production and not have some sort of safeguards like this.

Then there's also the issue of partial success, what if a command to one vMix instance succeeds while another fails? For some devices/apps it may be ideal to retry sending the command to the connection that failed, for some it's better not to attempt it again and instead raise an error to the user.

So while it 'technically' could be done in Companion, it would be bit of a hack job and not something I'd recommend as there would be too many pitfalls compared to doing it properly.

**Comment by kman1898** [comment]
You are assuming functionality though. All it needs to do is duplicate actions. Variables and feedback can still be treated exactly as you like and you can still have the connection for your module. It would just linked to the duplicate  specific action.

Any action you’re going to fully duplicate in the current companion would have the exact same overhead when using Millumin for example. Maybe in the example you have, vmix, it is different. I don’t use vmix so I can’t say. 

All the other checks you mention would be covered no differently than they already are. It would just take less steps to get there. Less duplication of actions same amount of variable and feedback creation.

**Comment by dnmeid** [comment]
@kman1898 thedist gave you some great insights of the difficulties implementing this request. It is not impossible for sure, but there are a lot of things to consider. Be assured that all of the devs would be more than happy if there was an easy solution and we really would like to offer that functionality, but given that it is not easy and there are a lot of other projects to be done with a higher priority, this request will need to wait in line.
Please feel free to develop it yourself and send us a pull request or sponsor a developer from the money you save with the reduced time effort.

**Comment by sweetapollo** [comment]
I would be happy to put forwards some money for this to be brought through.
